x/merkledb/intermediate_node_db.go

```diff
@@ -108,17 +108,38 @@ func (db *intermediateNodeDB) Get(key path) (*node, error) {
 	}
 	db.metrics.IntermediateNodeCacheMiss()
 
-	prefixedKey := addPrefixToKey(db.bufferPool, intermediateNodePrefix, key.Bytes())
+	dbKey := db.constructDBKey(key)
 	db.metrics.DatabaseNodeRead()
-	nodeBytes, err := db.baseDB.Get(prefixedKey)
+	nodeBytes, err := db.baseDB.Get(dbKey)
 	if err != nil {
 		return nil, err
 	}
-	db.bufferPool.Put(prefixedKey)
+	db.bufferPool.Put(dbKey)
 
 	return parseNode(key, nodeBytes)
 }
 
+func (db *intermediateNodeDB) constructDBKey(key path) []byte {
+	// We need differentiate between two paths of equal byte length but different token length
+	// so add the modulo remainder as a key suffix
+	dbKey := db.bufferPool.Get().([]byte)
+	defer db.bufferPool.Put(dbKey)
+	compressedKey := key.Serialize().Value
+	keyLen := len(compressedKey) + 1
+	if cap(dbKey) >= keyLen {
+		// The [] byte we got from the pool is big enough to hold the prefixed key
+		dbKey = dbKey[:keyLen]
+	} else {
+		// The []byte from the pool wasn't big enough.
+		// Put it back and allocate a new, bigger one
+		db.bufferPool.Put(dbKey)
+		dbKey = make([]byte, keyLen)
+	}
+	copy(dbKey, compressedKey)
+	dbKey[len(compressedKey)] = byte(len(key) % 2)
+	return addPrefixToKey(db.bufferPool, intermediateNodePrefix, dbKey)
+}
```

Comment:
mega nit - but I thought that we were going to map:
`[]byte{0x01, 0x02, 0x03}` to `[]byte{0x12, 0x38}` rather than `[]byte{0x12, 0x30, 0x01}`. I know this would only shave half a byte off each key on average - but seems nicer to me 🤷

Reply:
should be able to use `bits.TrailingZeros8` on the last byte to calculate the how many bits to drop